crates/flux-infer/src/refine_tree.rs

```diff
@@ -780,3 +791,328 @@ impl RefineCtxtTrace {
+            .collect();
+
+        // while there is a candidate constraint, that has NO BOT kvars in lhs
+        while let Some(cid) = candidates.pop() {
+            if let VertexId::KVar(kvid) = graph.head(cid) {
+                // un-BOT the head kvar
+                assignment.remove(kvid);
+                // remove the head kvar from all (bot) assumptions where it currently occurs
+                for cid in graph.cstrs_with_kvar_as_assumption(&kvid) {
+                    // if cid HEAD is a kvar
+                    if let VertexId::KVar(rhs_kvid) = graph.head(cid) {
+                        let Some(assms) = bot_assms.get_mut(&cid) else {
+                            tracked_span_bug!("missing constraint info for {:?}", cid);
+                        };
+                        assms.remove(&kvid);
+                        if assignment.contains(rhs_kvid) && assms.is_empty() {
+                            candidates.push(cid);
+                        }
+                    };
+                }
+            }
+        }
+
+        assignment
+    }
+
+    /// Computes the set of all kvars that can be assigned to Top (True),
+    /// because they do not (transitively) reach any concrete HEAD.
+    fn top_kvars(&self, graph: Graph) -> Assignment {
+        // initialize
+        let mut assignment = Assignment::new(graph.kvars(), Label::Top);
+
+        // set of kvar {k | cid in graph.edges, c.rhs is concrete, k in c.lhs }
+        let mut candidates = vec![];
+        for info in graph.edges.values() {
+            if matches!(info.rhs, VertexId::Conc) {
+                for kvid in &info.lhs {
+                    candidates.push(*kvid);
+                }
+            }
+        }
+
+        // set each kvar that transitively reaches a concrete HEAD to NON-BOT
+        while let Some(kvid) = candidates.pop() {
+            // set that kvar to non-top
+            assignment.remove(kvid);
+
+            // for each constraint where kvid appears as head
+            for cid in graph.kv_rhs.get(&kvid).unwrap_or(&FxHashSet::default()) {
+                let info = graph
+                    .edges
+                    .get(cid)
+                    .unwrap_or_else(|| tracked_span_bug!("missing constraint info for {:?}", cid));
+                // add kvars in lhs to candidates (if they have not already been solved to non-BOT)
+                for lhs_kvid in &info.lhs {
+                    if assignment.contains(*lhs_kvid) {
+                        candidates.push(*lhs_kvid);
+                    }
+                }
+            }
+        }
+
+        assignment
+    }
+}
+
+#[derive(Debug)]
+struct Graph {
+    /// description of each constraint
+    edges: FxHashMap<CstrId, CstrInfo>,
+    /// set of edges where kvid appears as ASSM
+    kv_lhs: FxHashMap<KVid, FxHashSet<CstrId>>,
+    /// set of edges where kvid appears as HEAD
+    kv_rhs: FxHashMap<KVid, FxHashSet<CstrId>>,
```

Comment:
I think these two can be `Vec` because we are always inserting a fresh `CstrId`

```suggestion
    kv_lhs: FxHashMap<KVid, Vec<CstrId>>,
    /// set of edges where kvid appears as HEAD
    kv_rhs: FxHashMap<KVid, Vec<CstrId>>,
```

Reply:
Not quite, if the same KVar appears twice in an LHS then you'd end up inserting the CstrId twice...

Reply:
Good point, but you are already computing the `lhs` as a set just before pushing, so you can iterate over the deduplicated set of kvars.

Reply:
OTOH if you made the `ctx` a set then can make this a `Vec` ...

Reply:
but that makes it tedious to "pop" stuff off the context... leaving this for now.